src/middleware/cors/service.rs

```diff
@@ -1,41 +1,211 @@
-use super::Config;
+use super::{Config, CorsResource};
 
-use futures::{Poll};
-use http;
+use futures::{Async, Future, Poll};
+use http::{self, HeaderMap, StatusCode};
+use http::{Request, Response};
 use tower_service::Service;
 
 use std::sync::Arc;
 
+#[derive(Debug)]
 pub struct CorsService<S> {
     inner: S,
     config: Arc<Config>,
 }
 
 impl<S> CorsService<S> {
     pub(super) fn new(inner: S, config: Arc<Config>) -> CorsService<S> {
-        CorsService {
-            inner,
-            config,
-        }
+        CorsService { inner, config }
     }
 }
 
-use util::http::HttpService;
-use http::{Request, Response};
-
 impl<S> Service for CorsService<S>
-where S: HttpService,
+where
+    S: ::util::http::HttpService,
 {
     type Request = Request<S::RequestBody>;
-    type Response = Response<S::ResponseBody>;
+    type Response = Response<Option<S::ResponseBody>>;
     type Error = S::Error;
-    type Future = S::Future;
+    type Future = CorsFuture<S::Future>;
 
     fn poll_ready(&mut self) -> Poll<(), Self::Error> {
         self.inner.poll_ready()
     }
 
     fn call(&mut self, request: Self::Request) -> Self::Future {
-        self.inner.call(request)
+        match self.config.process_request(&request) {
+            Ok(CorsResource::Preflight(headers)) => CorsFuture::Handled(Some(headers)),
+            Ok(CorsResource::Simple(headers)) => {
+                CorsFuture::Simple(self.inner.call(request), Some(headers))
+            }
+            Err(e) => {
+                eprintln!("{:?}", e);
+                CorsFuture::Handled(None)
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum CorsFuture<F> {
```

Comment:
For backwards compatibility, I'd recommend making this a `struct CorsFuture`, with the inner enum private. That way, new variants can be added, and if we decided to change from `Option<HeaderMap>` to some other type with a more explicit name (for instance), we could.